File: td3learn/src/hexapod.cpp
#include <chrono>
#include <cmath>
#include <iostream>
#include <fstream>
#include <thread>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include "td3learn/hexapod.hpp"
#include "td3learn/utils.hpp"

// Include the user space hexapod interface
#include "hexapod.hpp"

namespace td3learn
{

    // HexapodEnvironment Implementation class
    class HexapodEnvironmentImpl
    {
    public:
        HexapodEnvironmentImpl(const HexapodEnvironmentConfig &config)
            : config(config),
              simulation_mode(config.type == "simulation"),
              last_action(config.action_dim),
              step_counter(0)
        {
            // Create hexapod instance
            hexapod = std::make_unique<hexapod::Hexapod>();

            // Initialize state and action dimensions based on robot configuration
            state_dim = 0;

            // IMU data: 3 accel + 3 gyro = 6 values
            if (config.use_imu)
            {
                state_dim += 6;
            }

            // Leg positions: 6 legs x 3 joints = 18 values
            if (config.use_leg_positions)
            {
                state_dim += hexapod::Config::NUM_LEGS * 3;
            }

            // Action dimension - one value per joint
            action_dim = hexapod::Config::NUM_LEGS * 3; // 6 legs x 3 joints

            utils::Logger::info("Created HexapodEnvironment with state_dim=" +
                                std::to_string(state_dim) + ", action_dim=" +
                                std::to_string(action_dim));
        }

        ~HexapodEnvironmentImpl()
        {
            // Clean up hexapod resources if initialized
            if (hexapod && initialized)
            {
                hexapod->cleanup();
            }
        }

        Result init()
        {
            try
            {
                // Initialize hexapod
                if (!hexapod->init())
                {
                    utils::Logger::error("Failed to initialize hexapod: " +
                                         hexapod->getLastErrorMessage());
                    return Result::ERROR_HARDWARE;
                }

                utils::Logger::info(simulation_mode ? "Connected to hexapod simulation" : "Connected to hexapod hardware");

                // Initialize state vector
                current_state.resize(state_dim);
                std::fill(current_state.begin(), current_state.end(), 0.0f);

                // Set action limits
                action_low.resize(action_dim);
                action_high.resize(action_dim);
                std::fill(action_low.begin(), action_low.end(), -1.0f); // Normalized to [-1, 1]
                std::fill(action_high.begin(), action_high.end(), 1.0f);

                last_action.resize(action_dim);
                std::fill(last_action.begin(), last_action.end(), 0.0f);

                step_counter = 0;
                initialized = true;
                return Result::SUCCESS;
            }
            catch (const std::exception &e)
            {
                utils::Logger::error("Exception initializing hexapod: " + std::string(e.what()));
                return Result::ERROR_INITIALIZATION;
            }
        }

        State reset()
        {
            step_counter = 0;

            // Center all legs
            if (!hexapod->centerAll())
            {
                utils::Logger::warning("Failed to center legs on reset: " +
                                       hexapod->getLastErrorMessage());
            }

            // Re-initialize state vector
            std::fill(current_state.begin(), current_state.end(), 0.0f);
            std::fill(last_action.begin(), last_action.end(), 0.0f);

            // Update state with current sensor data
            updateStateVector();

            return current_state;
        }

        Result step(const Action &action, State &next_state, Scalar &reward, bool &done, std::string &info)
        {
            if (action.size() != action_dim)
            {
                return Result::ERROR_INVALID_ARGUMENT;
            }

            // Apply actions to robot
            Result result = applyAction(action);
            if (result != Result::SUCCESS)
            {
                return result;
            }

            // Wait for step duration
            std::this_thread::sleep_for(std::chrono::milliseconds(
                static_cast<int>(config.step_time * 1000)));

            // Update state with new sensor readings
            updateStateVector();
            next_state = current_state;

            // Calculate reward
            reward = calculateReward(action);

            // Increment step counter and check if done
            step_counter++;
            done = (step_counter >= config.max_episode_steps);

            // Store action for next reward calculation
            last_action = action;

            info = ""; // No additional info for now
            return Result::SUCCESS;
        }

        int getStateDimension() const
        {
            return state_dim;
        }

        int getActionDimension() const
        {
            return action_dim;
        }

        void getActionLimits(State &low, State &high) const
        {
            low = action_low;
            high = action_high;
        }

        Result setSimulationMode(bool simulation)
        {
            if (simulation_mode == simulation)
            {
                return Result::SUCCESS; // No change needed
            }

            // Would need to reinitialize hexapod with different mode
            // This is a simplification - in real implementation, we would
            // recreate the hexapod with the new mode
            simulation_mode = simulation;
            return Result::SUCCESS;
        }

        bool isSimulation() const
        {
            return simulation_mode;
        }

    private:
        // Configuration
        HexapodEnvironmentConfig config;
        bool initialized = false;

        // Hexapod interface
        std::unique_ptr<hexapod::Hexapod> hexapod;

        // Simulation flag
        bool simulation_mode;

        // State and action information
        int state_dim;
        int action_dim;
        State current_state;
        Action action_low;
        Action action_high;
        Action last_action;

        // Episode tracking
        int step_counter;

        // Update the state vector from sensor data
        void updateStateVector()
        {
            int idx = 0;

            if (config.use_imu)
            {
                // Get IMU data
                hexapod::ImuData imu_data;

                if (!hexapod->getImuData(imu_data))
                {
                    utils::Logger::warning("Failed to read IMU data: " +
                                           hexapod->getLastErrorMessage());
                    // Use zeros for IMU data
                    imu_data = hexapod::ImuData();
                }

                // Normalize IMU data into state vector
                // Acceleration: ±2g range -> ±1.0 value
                current_state[idx++] = imu_data.getAccelX() / 2.0f;
                current_state[idx++] = imu_data.getAccelY() / 2.0f;
                current_state[idx++] = imu_data.getAccelZ() / 2.0f;

                // Gyroscope: ±250°/s range -> ±1.0 value
                current_state[idx++] = imu_data.getGyroX() / 250.0f;
                current_state[idx++] = imu_data.getGyroY() / 250.0f;
                current_state[idx++] = imu_data.getGyroZ() / 250.0f;
            }

            if (config.use_leg_positions)
            {
                // Get leg positions
                for (int leg = 0; leg < hexapod::Config::NUM_LEGS; leg++)
                {
                    hexapod::LegPosition leg_pos;

                    if (!hexapod->getLegPosition(leg, leg_pos))
                    {
                        utils::Logger::warning("Failed to get position for leg " +
                                               std::to_string(leg) + ": " +
                                               hexapod->getLastErrorMessage());
                        // Use zeros for leg position
                        leg_pos = hexapod::LegPosition();
                    }

                    // Normalize joint angles to [-1, 1] range
                    // Assuming 90 degrees is the max angle in each direction
                    current_state[idx++] = leg_pos.getHip() / 90.0f;
                    current_state[idx++] = leg_pos.getKnee() / 90.0f;
                    current_state[idx++] = leg_pos.getAnkle() / 90.0f;
                }
            }
        }

        // Apply an action to the robot
        Result applyAction(const Action &action)
        {
            // Check that action is the right size
            if (action.size() != action_dim)
            {
                return Result::ERROR_INVALID_ARGUMENT;
            }

            // Apply actions to each leg
            for (int leg = 0; leg < hexapod::Config::NUM_LEGS; leg++)
            {
                // Get joint angles from action vector
                // Convert normalized [-1, 1] to servo angles [-90, 90]
                hexapod::LegPosition leg_pos;
                leg_pos.leg_num = leg;

                int base_idx = leg * 3;
                leg_pos.setHip(static_cast<int16_t>(action[base_idx] * 90.0f));
                leg_pos.setKnee(static_cast<int16_t>(action[base_idx + 1] * 90.0f));
                leg_pos.setAnkle(static_cast<int16_t>(action[base_idx + 2] * 90.0f));

                // Apply to hardware
                if (!hexapod->setLegPosition(leg, leg_pos))
                {
                    utils::Logger::warning("Failed to set leg " + std::to_string(leg) +
                                           " position: " + hexapod->getLastErrorMessage());
                }
            }

            return Result::SUCCESS;
        }

        // Calculate reward based on current state and action
        Scalar calculateReward(const Action &action)
        {
            Scalar reward = 0.0f;

            // Forward velocity reward component - would come from IMU or external measurement
            // This is a simplification - in a real system, we'd need actual velocity feedback
            Scalar velocity_reward = 0.0f;
            if (config.use_imu)
            {
                // In a real system, we would calculate velocity from IMU data
                // This is a placeholder
                velocity_reward = 1.0f;
            }

            // Energy efficiency reward component - penalize large/wasteful movements
            Scalar energy_penalty = 0.0f;
            for (size_t i = 0; i < action.size(); i++)
            {
                // Quadratic penalty for action magnitudes
                energy_penalty += action[i] * action[i];
            }
            energy_penalty /= action.size();

            // Stability reward component - based on IMU data
            Scalar stability_reward = 0.0f;
            if (config.use_imu)
            {
                // In a real system, would calculate based on actual IMU readings
                stability_reward = 1.0f; // Placeholder
            }

            // Smoothness reward - penalize sudden changes in action
            Scalar smoothness_reward = 0.0f;
            if (last_action.size() > 0) // Replace last_action.empty() with size check
            {
                Scalar action_diff = 0.0f;
                for (size_t i = 0; i < action.size(); i++)
                {
                    Scalar diff = action[i] - last_action[i];
                    action_diff += diff * diff;
                }
                action_diff /= action.size();
                smoothness_reward = std::exp(-10.0f * action_diff); // High for small changes
            }

            // Combine rewards with weights
            reward = config.velocity_weight * velocity_reward -
                     config.energy_weight * energy_penalty +
                     config.stability_weight * stability_reward +
                     config.smoothness_weight * smoothness_reward;

            // Scale the final reward
            return reward * config.reward_scale;
        }
    };

    // HexapodEnvironment implementation methods
    HexapodEnvironment::HexapodEnvironment(const HexapodEnvironmentConfig &config)
        : Environment(config), impl(std::make_unique<HexapodEnvironmentImpl>(config)) {}

    HexapodEnvironment::~HexapodEnvironment() = default;

    Result HexapodEnvironment::init()
    {
        return impl->init();
    }

    State HexapodEnvironment::reset()
    {
        return impl->reset();
    }

    Result HexapodEnvironment::step(const Action &action, State &next_state,
                                    Scalar &reward, bool &done, std::string &info)
    {
        return impl->step(action, next_state, reward, done, info);
    }

    int HexapodEnvironment::getStateDimension() const
    {
        return impl->getStateDimension();
    }

    int HexapodEnvironment::getActionDimension() const
    {
        return impl->getActionDimension();
    }

    void HexapodEnvironment::getActionLimits(State &low, State &high) const
    {
        impl->getActionLimits(low, high);
    }

    Result HexapodEnvironment::setSimulationMode(bool simulation)
    {
        return impl->setSimulationMode(simulation);
    }

    bool HexapodEnvironment::isSimulation() const
    {
        return impl->isSimulation();
    }

} // namespace td3learn

File: pytd3/environment/hexapod.py
"""
Hexapod environment interface for reinforcement learning
"""
import time
import numpy as np
import gym
from gym import spaces

class HexapodEnvironment(gym.Env):
    """
    Hexapod environment implementing the OpenAI Gym interface.
    This environment can work with either hardware or simulation.
    """
    
    def __init__(
        self,
        use_hardware=False,
        action_dim=18,
        state_dim=24,
        max_episode_steps=1000,
        reward_weights=None,
        time_step=0.05
    ):
        """
        Initialize the hexapod environment.
        
        Args:
            use_hardware: Whether to use actual hardware or simulation
            action_dim: Dimension of the action space (typically 18 for 6 legs × 3 joints)
            state_dim: Dimension of the state space
            max_episode_steps: Maximum steps per episode
            reward_weights: Weights for different reward components
            time_step: Time step between actions (seconds)
        """
        super(HexapodEnvironment, self).__init__()
        
        self.use_hardware = use_hardware
        self.action_dim = action_dim
        self.state_dim = state_dim
        self.max_episode_steps = max_episode_steps
        self.time_step = time_step
        
        # Define default reward weights if none provided
        self.reward_weights = reward_weights or {
            'stability': 1.0,
            'energy': 0.5,
            'velocity': 0.8,
            'direction': 0.6,
            'smoothness': 0.4
        }
        
        # Define action and observation spaces
        self.action_space = spaces.Box(
            low=-1.0, high=1.0, shape=(action_dim,), dtype=np.float32
        )
        
        self.observation_space = spaces.Box(
            low=-np.inf, high=np.inf, shape=(state_dim,), dtype=np.float32
        )
        
        # Initialize state and episode counters
        self.current_state = np.zeros(state_dim, dtype=np.float32)
        self.steps = 0
        self.episode = 0
        self.last_action = np.zeros(action_dim, dtype=np.float32)
        
        # Initialize hardware interface if using hardware
        self.hardware = None
        if use_hardware:
            self._setup_hardware()
    
    def _setup_hardware(self):
        """Set up hardware interface"""
        try:
            from pytd3.bridge.client import HexapodHardwareClient
            self.hardware = HexapodHardwareClient()
            success = self.hardware.connect()
            if not success:
                print("WARNING: Failed to connect to hexapod hardware. Using simulation instead.")
                self.use_hardware = False
                self.hardware = None
        except ImportError as e:
            print(f"WARNING: Hardware interface not available: {e}")
            self.use_hardware = False
            self.hardware = None
    
    def reset(self):
        """
        Reset the environment to initial state.
        
        Returns:
            Initial observation
        """
        self.steps = 0
        self.episode += 1
        
        # Reset hardware if using actual hardware
        if self.use_hardware and self.hardware:
            self.hardware.reset()
        
        # Generate initial state
        self.current_state = self._get_observation()
        self.last_action = np.zeros(self.action_dim, dtype=np.float32)
        
        return self.current_state
    
    def step(self, action):
        """
        Take a step in the environment.
        
        Args:
            action: Action to take
            
        Returns:
            observation: New state observation
            reward: Reward for the action
            done: Whether the episode is finished
            info: Additional information
        """
        # Increment step counter
        self.steps += 1
        
        # Clip action to valid range
        action = np.clip(action, -1.0, 1.0)
        
        # Apply action to hardware or simulation
        if self.use_hardware and self.hardware:
            self.hardware.apply_action(action)
        else:
            # Simulate action in software
            self._simulate_action(action)
        
        # Save action for smoothness calculations
        self.last_action = action.copy()
        
        # Get new state after action
        self.current_state = self._get_observation()
        
        # Calculate reward
        reward = self._calculate_reward(action)
        
        # Check if episode is done
        done = self.steps >= self.max_episode_steps
        
        # Additional info
        info = {
            'steps': self.steps,
            'episode': self.episode
        }
        
        return self.current_state, reward, done, info
    
    def _get_observation(self):
        """
        Get current observation (state).
        
        Returns:
            Current state observation
        """
        if self.use_hardware and self.hardware:
            # Get state from hardware
            return self.hardware.get_state()
        else:
            # Generate simulated state
            return self._generate_simulated_state()
    
    def _generate_simulated_state(self):
        """
        Generate a simulated state for testing without hardware.
        
        Returns:
            Simulated state vector
        """
        # In a real implementation, this would simulate hexapod physics
        # Here we just create a simple state with some basic features
        
        # Simple simulation for testing
        state = np.zeros(self.state_dim, dtype=np.float32)
        
        # Add IMU-like data (orientation + acceleration)
        state[0:3] = np.random.normal(0, 0.1, 3)  # orientation
        state[3:6] = np.random.normal(0, 0.2, 3)  # acceleration
        
        # Add joint positions (similar to what the hardware would report)
        if hasattr(self, 'last_action') and self.last_action is not None:
            # Map actions to joint positions
            joint_positions = self.last_action.copy()
            state[6:6+self.action_dim] = joint_positions
        
        return state
    
    def _simulate_action(self, action):
        """
        Simulate the effect of an action without hardware.
        
        Args:
            action: Action to simulate
        """
        # In a real implementation, this would update a physics simulation
        # For now, we'll just add a delay to simulate execution time
        time.sleep(self.time_step)
    
    def _calculate_reward(self, action):
        """
        Calculate reward for the current state and action.
        
        Args:
            action: Action taken
            
        Returns:
            Total reward value
        """
        # Initialize reward components
        reward_components = {}
        
        if self.use_hardware and self.hardware:
            # Get reward components from hardware
            reward_components = self.hardware.get_reward_components()
        else:
            # Calculate simulated reward components
            # Stability reward (higher when orientation is level)
            stability = 1.0 - min(1.0, np.sum(np.abs(self.current_state[0:3])) / 3)
            
            # Energy efficiency (penalize high actuator values)
            energy = 1.0 - min(1.0, np.mean(np.abs(action)) / 0.7)
            
            # Velocity reward (simplified for simulation)
            velocity = max(0, np.mean(self.current_state[3:6]))
            
            # Direction reward (simplified)
            direction = 0.5  # Neutral value for simulation
            
            # Smoothness reward (penalize sudden changes)
            if hasattr(self, 'last_action') and self.last_action is not None:
                action_diff = np.abs(action - self.last_action)
                smoothness = 1.0 - min(1.0, np.mean(action_diff) * 5)
            else:
                smoothness = 0.5
            
            reward_components = {
                'stability': stability,
                'energy': energy,
                'velocity': velocity,
                'direction': direction,
                'smoothness': smoothness
            }
        
        # Calculate weighted sum of reward components
        total_reward = 0
        for component, value in reward_components.items():
            if component in self.reward_weights:
                total_reward += value * self.reward_weights[component]
        
        return total_reward
    
    def close(self):
        """Clean up resources"""
        if self.use_hardware and self.hardware:
            self.hardware.disconnect()
            self.hardware = None
    
    def render(self, mode='human'):
        """
        Render the environment.
        Not implemented for hexapod hardware.
        """
        pass